src/main/java/org/logstash/plugins/inputs/http/util/ExecutionObserver.java

```diff
@@ -0,0 +1,201 @@
+package org.logstash.plugins.inputs.http.util;
+
+import java.lang.invoke.MethodHandles;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.LongSupplier;
+
+/**
+ * An {@code ExecutionObserver} observes possibly-concurrent execution, and provides information about the
+ * longest-running observed execution.
+ *
+ * <p>
+ *     It is concurrency-safe and non-blocking, and uses plain memory access where practical.
+ * </p>
+ */
+public class ExecutionObserver {
+    private final AtomicReference<Execution> head; // newest execution
+    private final AtomicReference<Execution> tail; // oldest execution
```

Comment:
TODO: invert to match java semantics, where things go into the tail (back) of a queue and are removed from the head (front)